fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/synchronizedsession/SynchronizedSessionRunner.java

```diff
@@ -74,20 +107,46 @@ public static SynchronizedSessionRunner joinSession(@Nonnull Subspace lockSubspa
         return new SynchronizedSessionRunner(runner, session);
     }
 
-    private SynchronizedSessionRunner(FDBDatabaseRunner underlyingRunner, SynchronizedSession session) {
+    private SynchronizedSessionRunner(@Nonnull FDBDatabaseRunner underlyingRunner,
+                                      @Nonnull SynchronizedSession session) {
         this.underlying = underlyingRunner;
         this.session = session;
     }
 
+    // Check and renew the lock when the session in being used.
+    // TODO: Maybe the time should be updated even if the work is failed. For example, online indexer may fail
+    // in the first a few transactions to find the optimal number of records to scan in one transaction.
+    private <T> Function<FDBRecordContext, T> runInSession(
+            @Nonnull Function<? super FDBRecordContext, ? extends T> work) {
+        return context -> {
+            context.asyncToSync(FDBStoreTimer.Waits.WAIT_CHECK_SYNC_SESSION, session.checkLockAsync((context.ensureActive())));
+            T result = work.apply(context);
+            session.updateLockSessionLeaseEndTime(context.ensureActive());
+            return result;
+        };
+    }
+
+    private <T> Function<? super FDBRecordContext, CompletableFuture<? extends T>> runInSessionAsync(
+            @Nonnull Function<? super FDBRecordContext, CompletableFuture<? extends T>> work) {
+        return context -> session.checkLockAsync(context.ensureActive())
+                .thenCompose(vignore -> work.apply(context))
+                .thenApply(result -> {
+                    session.updateLockSessionLeaseEndTime(context.ensureActive());
+                    return result;
+                });
+    }
+
     public UUID getSessionId() {
-        return this.session.getSessionId();
+        return session.getSessionId();
     }
 
-    // This is not necessarily to be called when closing the runner because there can be multiple runner for a same
-    // session.
-    public void closeSession() {
+    /**
+     * Releases the lock to end the synchronized session. This should not necessarily be called when closing the runner
+     * because there can be multiple runners for the same session.
+     */
+    public void endSession() {
```

Comment:
Should there be an `endSessionAsync` that calls `underlying.runAsync`?

I think the difference is just that `underlying.run` will block on the commit, which some people may want to avoid.